<a href="https://colab.research.google.com/github/cristiandarioortegayubro/UNI/blob/main/colab/dc_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

![](https://github.com/cristiandarioortegayubro/UNI/blob/main/UNI%2039%20Aniversario.jpg?raw=true)

# **Limpieza de datos**

Nos encontramos iniciando un nuevo proyecto, nuestra empresa es una pequeña inmobiliaria que quiere fusionar el potencial del manejo de datos con la venta/alquiler de propiedades. Al ser una pequeña empresa el ***departamento de datos*** esta compuesto por una persona: nosotros. 

Esto quiere decir que vamos a absorber los roles de:

- Ingeniero de Datos, y 
- Analista de Datos.


## **Ingesta de datos de propiedades**

***El dataset central de este proyecto de propiedades de Argentina, tiene la siguientes columnas:***

- **id** - Identificador del aviso. No es único: si el aviso es actualizado por la inmobiliaria (nueva versión del aviso) se crea un nuevo registro con la misma id pero distintas fechas: de alta y de baja.
- **ad_type** - Tipo de aviso (Propiedad, Desarrollo/Proyecto).
- **start_date** - Fecha de alta del aviso.
- **end_date** - Fecha de baja del aviso.
- **created_on** - Fecha de alta de la primera versión del aviso.
- **lat** - Latitud.
- **lon** - Longitud.
- **l1** - Nivel administrativo 1: país.
- **l2** - Nivel administrativo 2: usualmente provincia.
- **l3** - Nivel administrativo 3: usualmente ciudad.
- **l4** - Nivel administrativo 4: usualmente barrio.
- **l5** - sin informacion
- **l6** - sin informacion
- **rooms** - Cantidad de ambientes
- **bedrooms** - Cantidad de dormitorios
- **bathrooms** - Cantidad de baños.
- **surface_total** - Superficie total en m².
- **surface_covered** - Superficie cubierta en m².
- **price** - Precio publicado en el anuncio.
- **currency** - Moneda del precio publicado.
- **price_period** - Periodo del precio (Diario, Semanal, Mensual)
- **title** - Título del anuncio.
- **description** - Descripción del anuncio.
- **property_type**: - Caracteristicas de la propiedad
- **operation_type**: - Caracteristicas de la operacion ['Venta', 'Alquiler', 'Alquiler temporal']

## **Habilitando bibliotecas**

Habilitar las bibliotecas necesarias para generar un dataframe, hacer limpieza de datos, y generar gráficos.

### **Instalando bibliotecas necesarias**

### **Para analisis de datos**

In [1]:
import pandas as pd

### **Para limpieza de datos**

In [2]:
import numpy as np

### **Para gráficos**

In [3]:
import seaborn as sns
import plotly.express as px

## **Análisis Exploratorio de los Datos**

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


Leer el archivo  ```.csv``` con Pandas y guardarlo en una variable llamada ```prop```

In [5]:
#prop = pd.read_csv("/content/drive/MyDrive/Notebooks/Workshop/propiedades.csv")

***Observar los primeros 8 registros***

In [6]:
prop.head(8)

id    ad_type  start_date    end_date  created_on  \
0  DyVXfkpKygVBKuUk5olH+A==  Propiedad  2020-08-22  2020-09-03  2020-08-22   
1  9naojilaMecJN4jlQiTkGg==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
2  tlCPRJPjoDEUzuuCelemAQ==  Propiedad  2020-08-22  2020-08-31  2020-08-22   
3  Zw3b91glQUO3HNrM5fPYlQ==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
4  bsU81gm9JEgtZCbTYgvykg==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
5  3Cips/GEPuJqktYycSALCQ==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
6  6TO5dWwmivuJ67zkztUzCA==  Propiedad  2020-08-22  2020-09-04  2020-08-22   
7  VlE71iojfIFoH4piJfFQ/g==  Propiedad  2020-08-22  2020-09-04  2020-08-22   

         lat        lon         l1                            l2  \
0 -34.407468 -58.957367  Argentina      Bs.As. G.B.A. Zona Norte   
1 -37.996039 -57.542509  Argentina  Buenos Aires Costa Atlántica   
2 -31.380187 -58.009182  Argentina                    Entre Ríos   
3 -27.494106 -55.123455  Argentina                      Misiones   
4 -32.948856 -60.630464  Argentina                      Santa Fe   
5 -32.884278 -60.710901  Argentina                      Santa Fe   
6 -34.604195 -58.559827  Argentina      Bs.As. G.B.A. Zona Oeste   
7 -32.948834 -60.630169  Argentina                      Santa Fe   

                l3  ... bathrooms surface_total  surface_covered  price  \
0            Pilar  ...       NaN      133139.0              NaN    NaN   
1    Mar del Plata  ...       NaN         687.0            687.0    NaN   
2        Concordia  ...       1.0          80.0             80.0    NaN   
3            Oberá  ...       1.0           NaN              NaN    NaN   
4          Rosario  ...       1.0          76.0             66.0    NaN   
5          Rosario  ...       1.0          39.0             33.0    NaN   
6  Tres de Febrero  ...       1.0           NaN             42.0    NaN   
7          Rosario  ...       2.0         130.0            108.0    NaN   

   currency  price_period                                              title  \
0       NaN       Mensual  VENTA - Lote Industrial 130.000m2 - Parque Ind...   
1       NaN       Mensual                               Casa - Santa Cecilia   
2       NaN       Mensual                                   Casa - Concordia   
3       NaN       Mensual                                    Terreno - Obera   
4       NaN       Mensual                             Departamento - Rosario   
5       NaN       Mensual                               9 DE JULIO 535 06 01   
6       NaN       Mensual                             Departamento - Caseros   
7       NaN       Mensual                             Departamento - Rosario   

                                         description  property_type  \
0  Lote (Nro.50) mide 133.139 m2\n<br>Valor: U$S ...           Lote   
1  Conjunto edilicio de casa y departamentos en b...           Otro   
2  Casa de un dormitorio, cocina, baño, living co...           Casa   
3  IMPORTANTE PROPIEDAD EN VENTA EN OBERA MISIONE...           Lote   
4  PORTAL DE LOS MARINOS<br>Es un complejo de viv...   Departamento   
5  CARACTERÍSTICAS<br>Ingreso living comedor con ...   Departamento   
6  Departamento 2 ambientes con Lavadero en Nicar...   Departamento   
7  PORTAL DE LOS MARINOS<br>Es un complejo de viv...   Departamento   

  operation_type  
0          Venta  
1          Venta  
2       Alquiler  
3          Venta  
4          Venta  
5       Alquiler  
6       Alquiler  
7          Venta  

[8 rows x 25 columns]

***Observar los últimos 5 registros***

In [7]:
prop.tail()

id    ad_type  start_date    end_date  \
999995  jMYhwydKqUje8HCx7RzwJQ==  Propiedad  2020-12-04  9999-12-31   
999996  FsP7ornzqnK1CvUL8quQ9Q==  Propiedad  2020-12-04  9999-12-31   
999997  fJLVUIziPl31ACoaE/fj1g==  Propiedad  2020-12-04  9999-12-31   
999998  7mW3bUSJFA0RBOj2xtBi+w==  Propiedad  2020-12-04  9999-12-31   
999999  rEynGyGrrsxNeG0ib5vUBA==  Propiedad  2020-12-04  2020-12-21   

        created_on        lat        lon         l1                        l2  \
999995  2020-12-04 -32.958863 -60.665633  Argentina                  Santa Fe   
999996  2020-12-04 -34.596587 -58.376893  Argentina           Capital Federal   
999997  2020-12-04 -34.576445 -58.432603  Argentina           Capital Federal   
999998  2020-12-04 -32.866636 -68.841144  Argentina                   Mendoza   
999999  2020-12-04 -34.447524 -58.631552  Argentina  Bs.As. G.B.A. Zona Norte   

             l3  ... bathrooms surface_total  surface_covered     price  \
999995  Rosario  ...       7.0         400.0            300.0  400000.0   
999996   Retiro  ...       5.0         465.0            465.0    3700.0   
999997  Palermo  ...       2.0         615.0            425.0  570000.0   
999998      NaN  ...      20.0         450.0            450.0  800000.0   
999999    Tigre  ...       4.0         350.0            300.0    7000.0   

        currency  price_period  \
999995       USD       Mensual   
999996       USD       Mensual   
999997       USD       Mensual   
999998       USD       Mensual   
999999       USD       Mensual   

                                                    title  \
999995  Se vende imponente casa a minutos del centro d...   
999996                        Edificio Comercial - Retiro   
999997  Edificio Comercial  en excelente zona de Paler...   
999998                       Edificio Comercial - Capital   
999999  INCREÍBLE ECASA A LA LAGUNA DESDE DIC a FEBRER...   

                                              description    property_type  \
999995  Esta excelente propiedad se encuentra ubicada ...             Casa   
999996  GRAN EDIFICIO PETIT HOTEL ESTILO FRANCÉS  CON ...  Local comercial   
999997  Edificio Comercial CON ENTRADA INDEPENDIENTE S...  Local comercial   
999998  Local 450m² con Cloaca en 25 de mayo, Argentin...  Local comercial   
999999  ESPECTACULAR CASA A LA LAGUNA  -BARRIO SANTA B...             Casa   

           operation_type  
999995              Venta  
999996           Alquiler  
999997              Venta  
999998              Venta  
999999  Alquiler temporal  

[5 rows x 25 columns]

***Imprimir la cantidad de filas y columnas del dataframe***

In [8]:
prop.shape
fil = prop.shape[0]
col = prop.shape[1]

In [9]:
print(f"El dataframe tiene {fil} filas y {col} columnas")

El dataframe tiene 1000000 filas y 25 columnas


***Usar info***

In [10]:
prop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 25 columns):
 #   Column           Non-Null Count    Dtype  
---  ------           --------------    -----  
 0   id               1000000 non-null  object 
 1   ad_type          1000000 non-null  object 
 2   start_date       1000000 non-null  object 
 3   end_date         1000000 non-null  object 
 4   created_on       1000000 non-null  object 
 5   lat              894233 non-null   float64
 6   lon              894191 non-null   float64
 7   l1               1000000 non-null  object 
 8   l2               1000000 non-null  object 
 9   l3               965273 non-null   object 
 10  l4               306162 non-null   object 
 11  l5               5530 non-null     object 
 12  l6               0 non-null        float64
 13  rooms            714179 non-null   float64
 14  bedrooms         649933 non-null   float64
 15  bathrooms        765122 non-null   float64
 16  surface_total    47

***Usar describe para variables cuantitativas y mostrar el resumen con dos decimales y transpuesto (transposicion)***

In [11]:
round(prop.describe(), 2).T

count       mean         std     min       25%       50%  \
lat              894233.0     -34.38        3.04  -54.84    -34.72    -34.59   
lon              894191.0     -59.49        2.80 -122.54    -58.88    -58.49   
l6                    0.0        NaN         NaN     NaN       NaN       NaN   
rooms            714179.0       2.92        1.74    1.00      2.00      3.00   
bedrooms         649933.0       2.10        2.08  -16.00      1.00      2.00   
bathrooms        765122.0       1.71        1.09    1.00      1.00      1.00   
surface_total    477831.0     513.76     4297.24 -136.00     50.00     95.00   
surface_covered  487756.0   11656.22  4443191.26 -130.00     45.00     78.00   
price            958243.0  346094.02  5713689.88    0.00  38000.00  97000.00   

                       75%           max  
lat                 -34.42  4.963000e+01  
lon                 -58.40  1.800000e+02  
l6                     NaN           NaN  
rooms                 4.00  4.000000e+01  
bedrooms              3.00  9.000000e+02  
bathrooms             2.00  2.000000e+01  
surface_total       250.00  2.000000e+05  
surface_covered     168.00  2.147484e+09  
price            210000.00  3.100000e+09

In [12]:
round(prop.describe().transpose(), 2)

count       mean         std     min       25%       50%  \
lat              894233.0     -34.38        3.04  -54.84    -34.72    -34.59   
lon              894191.0     -59.49        2.80 -122.54    -58.88    -58.49   
l6                    0.0        NaN         NaN     NaN       NaN       NaN   
rooms            714179.0       2.92        1.74    1.00      2.00      3.00   
bedrooms         649933.0       2.10        2.08  -16.00      1.00      2.00   
bathrooms        765122.0       1.71        1.09    1.00      1.00      1.00   
surface_total    477831.0     513.76     4297.24 -136.00     50.00     95.00   
surface_covered  487756.0   11656.22  4443191.26 -130.00     45.00     78.00   
price            958243.0  346094.02  5713689.88    0.00  38000.00  97000.00   

                       75%           max  
lat                 -34.42  4.963000e+01  
lon                 -58.40  1.800000e+02  
l6                     NaN           NaN  
rooms                 4.00  4.000000e+01  
bedrooms              3.00  9.000000e+02  
bathrooms             2.00  2.000000e+01  
surface_total       250.00  2.000000e+05  
surface_covered     168.00  2.147484e+09  
price            210000.00  3.100000e+09

***Usar describe para variables cualitativas y mostrar resumen transpuesto***

In [13]:
prop.describe(include=[object]).transpose()

count   unique  \
id              1000000  1000000   
ad_type         1000000        1   
start_date      1000000      362   
end_date        1000000      450   
created_on      1000000      362   
l1              1000000        4   
l2              1000000       43   
l3               965273     1367   
l4               306162     1070   
l5                 5530       21   
currency         955491        5   
price_period     429870        3   
title            999999   548833   
description      999958   817449   
property_type   1000000       10   
operation_type  1000000        3   

                                                           top     freq  
id                                    DyVXfkpKygVBKuUk5olH+A==        1  
ad_type                                              Propiedad  1000000  
start_date                                          2020-10-02    25159  
end_date                                            9999-12-31   208960  
created_on                                          2020-10-02    25159  
l1                                                   Argentina   983115  
l2                                             Capital Federal   265125  
l3                                                     Rosario    61746  
l4                                                    Nordelta    19479  
l5                                            BarrioPortezuelo      735  
currency                                                   USD   738157  
price_period                                           Mensual   429842  
title                                               casa-venta     3338  
description     Departamento en Alquiler de 1 dorm. en Cordoba     1980  
property_type                                     Departamento   457907  
operation_type                                           Venta   750831

In [14]:
prop.describe(include=[object]).T

count   unique  \
id              1000000  1000000   
ad_type         1000000        1   
start_date      1000000      362   
end_date        1000000      450   
created_on      1000000      362   
l1              1000000        4   
l2              1000000       43   
l3               965273     1367   
l4               306162     1070   
l5                 5530       21   
currency         955491        5   
price_period     429870        3   
title            999999   548833   
description      999958   817449   
property_type   1000000       10   
operation_type  1000000        3   

                                                           top     freq  
id                                    DyVXfkpKygVBKuUk5olH+A==        1  
ad_type                                              Propiedad  1000000  
start_date                                          2020-10-02    25159  
end_date                                            9999-12-31   208960  
created_on                                          2020-10-02    25159  
l1                                                   Argentina   983115  
l2                                             Capital Federal   265125  
l3                                                     Rosario    61746  
l4                                                    Nordelta    19479  
l5                                            BarrioPortezuelo      735  
currency                                                   USD   738157  
price_period                                           Mensual   429842  
title                                               casa-venta     3338  
description     Departamento en Alquiler de 1 dorm. en Cordoba     1980  
property_type                                     Departamento   457907  
operation_type                                           Venta   750831

In [15]:
prop.describe(include=["object"]).T

count   unique  \
id              1000000  1000000   
ad_type         1000000        1   
start_date      1000000      362   
end_date        1000000      450   
created_on      1000000      362   
l1              1000000        4   
l2              1000000       43   
l3               965273     1367   
l4               306162     1070   
l5                 5530       21   
currency         955491        5   
price_period     429870        3   
title            999999   548833   
description      999958   817449   
property_type   1000000       10   
operation_type  1000000        3   

                                                           top     freq  
id                                    DyVXfkpKygVBKuUk5olH+A==        1  
ad_type                                              Propiedad  1000000  
start_date                                          2020-10-02    25159  
end_date                                            9999-12-31   208960  
created_on                                          2020-10-02    25159  
l1                                                   Argentina   983115  
l2                                             Capital Federal   265125  
l3                                                     Rosario    61746  
l4                                                    Nordelta    19479  
l5                                            BarrioPortezuelo      735  
currency                                                   USD   738157  
price_period                                           Mensual   429842  
title                                               casa-venta     3338  
description     Departamento en Alquiler de 1 dorm. en Cordoba     1980  
property_type                                     Departamento   457907  
operation_type                                           Venta   750831

***Ver si existen valores nulos***

In [16]:
prop.isna().sum().sort_values(ascending = False)

l6                 1000000
l5                  994470
l4                  693838
price_period        570130
surface_total       522169
surface_covered     512244
bedrooms            350067
rooms               285821
bathrooms           234878
lon                 105809
lat                 105767
currency             44509
price                41757
l3                   34727
description             42
title                    1
property_type            0
id                       0
ad_type                  0
l2                       0
l1                       0
created_on               0
end_date                 0
start_date               0
operation_type           0
dtype: int64

***Ver si existen valores duplicados***

In [17]:
prop.duplicated().sum()

0

## **Limpieza de los Datos**

***Ver si los nombres de las columnas tienen algun espacio oculto que nos dificulte al tratar de invocarlas, si encontras espacios ocultos eliminar esos espacios ocultos.***

In [18]:
prop.columns

Index(['id', 'ad_type', 'start_date', 'end_date', 'created_on', 'lat', 'lon',
       'l1', 'l2', 'l3', 'l4', 'l5', 'l6', 'rooms', 'bedrooms', 'bathrooms',
       'surface_total', 'surface_covered', 'price', 'currency', 'price_period',
       'title', 'description', 'property_type', 'operation_type'],
      dtype='object')

In [19]:
list(prop.columns)

['id',
 'ad_type',
 'start_date',
 'end_date',
 'created_on',
 'lat',
 'lon',
 'l1',
 'l2',
 'l3',
 'l4',
 'l5',
 'l6',
 'rooms',
 'bedrooms',
 'bathrooms',
 'surface_total',
 'surface_covered',
 'price',
 'currency',
 'price_period',
 'title',
 'description',
 'property_type',
 'operation_type']

***Generar un dataframe o una serie de pandas con el porcentaje de nulos (nan) que tiene cada columna, ordenado del mayor al menor.***

In [20]:
(prop.isna().sum()*100/prop.shape[0]).sort_values(ascending = False)

l6                 100.0000
l5                  99.4470
l4                  69.3838
price_period        57.0130
surface_total       52.2169
surface_covered     51.2244
bedrooms            35.0067
rooms               28.5821
bathrooms           23.4878
lon                 10.5809
lat                 10.5767
currency             4.4509
price                4.1757
l3                   3.4727
description          0.0042
title                0.0001
property_type        0.0000
id                   0.0000
ad_type              0.0000
l2                   0.0000
l1                   0.0000
created_on           0.0000
end_date             0.0000
start_date           0.0000
operation_type       0.0000
dtype: float64

In [21]:
round((prop.isna().sum()*100/prop.shape[0]).sort_values(ascending = False),2)

l6                 100.00
l5                  99.45
l4                  69.38
price_period        57.01
surface_total       52.22
surface_covered     51.22
bedrooms            35.01
rooms               28.58
bathrooms           23.49
lon                 10.58
lat                 10.58
currency             4.45
price                4.18
l3                   3.47
description          0.00
title                0.00
property_type        0.00
id                   0.00
ad_type              0.00
l2                   0.00
l1                   0.00
created_on           0.00
end_date             0.00
start_date           0.00
operation_type       0.00
dtype: float64

***Eliminar columnas que tengan mas del $30$% de nulos (nans), luego de este paso deberian quedar $18$ columnas si lo hiciste bien.***

In [ ]:
prop.shape

In [22]:
prop.dropna(axis = 1, inplace = True, thresh = int(prop.shape[0] * 0.7))

In [23]:
prop.shape

(1000000, 18)

***Quedarse con solo los registros que tienen un $90$% de sus celdas libres de nulos (nans), luego de este paso deberia quedarte $922772$ registros.***

In [24]:
prop.shape

(1000000, 18)

In [26]:
prop.dropna(axis=0, inplace= True, thresh = int(prop.shape[1] * 0.9))

In [27]:
prop.shape

(922772, 18)

***El analisis se realizará sobre el *AMBA*, eliminar todo lo que no sea parte del mismo. Luego de este paso deberian quedarte $617747$ registros.***

AMBA es Area Metropolitana de Buenos Aires y comprende:
+ Bs. As. Zona Oeste
+ Bs. As. Zona Norte
+ Bs. As. Zona Sur
+ Capital Federal

In [28]:
prop['l2'].unique()

array(['Entre Ríos', 'Misiones', 'Santa Fe', 'Bs.As. G.B.A. Zona Oeste',
       'Bs.As. G.B.A. Zona Norte', 'Tucumán',
       'Buenos Aires Costa Atlántica', 'Chubut', 'Capital Federal',
       'Córdoba', 'Buenos Aires Interior', 'Bs.As. G.B.A. Zona Sur',
       'San Luis', 'Neuquén', 'Colonia', 'Río Negro', 'Montevideo',
       'Maldonado', 'Mendoza', 'Corrientes', 'Salta', 'Florida',
       'Tierra Del Fuego', 'Chaco', 'Santa Cruz', 'La Pampa', 'Catamarca',
       'Canelones', 'Santiago Del Estero', 'San Juan', 'Miami', 'Jujuy',
       'Michigan', 'São Paulo', 'La Rioja', 'Rocha', 'Santa Catarina',
       'Pennsylvania', 'Rio Grande do Norte', 'Maryland', 'Formosa',
       'Rio de Janeiro', 'California'], dtype=object)

In [29]:
prop = prop[(prop['l2'] == 'Capital Federal') | 
            (prop['l2'] == 'Bs.As. G.B.A. Zona Norte') | 
            (prop['l2'] == 'Bs.As. G.B.A. Zona Oeste') | 
            (prop['l2'] == 'Bs.As. G.B.A. Zona Sur')]

In [30]:
prop.shape

(617747, 18)

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (617747, 18)**

***Chequear la columna l1 que representa al pais, deberia existir solo un valor: Argentina. Si hay otros valores eliminarlos***

In [31]:
prop['l1'].unique()

array(['Argentina'], dtype=object)

***Los registros que tienen el mismo id son de la misma propiedad pero actualizado, quedarse con el mas actualizado, los demas eliminarlos.***

In [32]:
(prop['id'].duplicated()).value_counts()

False    617747
Name: id, dtype: int64

***Eliminar las columnas:***

- id porque ya sabemos que todos los registros son diferentes y no vamos a usar esta columna para identificarlos.
- l1 porque solo tiene un valor: Argentina. No aporta informacion a los datos.
- description y title porque no contamos con herramientas de NLP para volverlas informacion para visualizar o modelizar.


In [33]:
prop.drop(['id', 'l1', 'description', 'title'], axis=1, inplace=True)

***Ver los primeros registros del dataframe.***

In [34]:
prop.head()

ad_type  start_date    end_date  created_on        lat        lon  \
6   Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.604195 -58.559827   
10  Propiedad  2020-08-22  2020-08-25  2020-08-22 -34.439610 -58.629968   
16  Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.633053 -58.419906   
17  Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.397401 -58.649185   
18  Propiedad  2020-08-22  2020-09-04  2020-08-22 -34.397401 -58.649185   

                          l2               l3  rooms  bathrooms    price  \
6   Bs.As. G.B.A. Zona Oeste  Tres de Febrero    2.0        1.0      NaN   
10  Bs.As. G.B.A. Zona Norte            Tigre    4.0        4.0      NaN   
16           Capital Federal            Boedo    NaN        NaN  70000.0   
17  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  35000.0   
18  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  27000.0   

   currency property_type     operation_type  
6       NaN  Departamento           Alquiler  
10      NaN          Casa  Alquiler temporal  
16      ARS       Oficina           Alquiler  
17      ARS       Oficina           Alquiler  
18      ARS       Oficina           Alquiler

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (617747, 14)**

In [35]:
prop.shape

(617747, 14)

***Sospechamos que las columnas start_date y created_on en este caso tienen los mismos valores, chequearlo, si son iguales eliminar la columna created_on***

In [36]:
(prop['start_date'] == prop['created_on']).value_counts()

True    617747
dtype: int64

In [37]:
prop.drop(['created_on'], axis=1, inplace=True)

In [38]:
prop.shape

(617747, 13)

***Ver los primeros registros del dataframe.***

In [39]:
prop.head()

ad_type  start_date    end_date        lat        lon  \
6   Propiedad  2020-08-22  2020-09-04 -34.604195 -58.559827   
10  Propiedad  2020-08-22  2020-08-25 -34.439610 -58.629968   
16  Propiedad  2020-08-22  2020-09-04 -34.633053 -58.419906   
17  Propiedad  2020-08-22  2020-09-04 -34.397401 -58.649185   
18  Propiedad  2020-08-22  2020-09-04 -34.397401 -58.649185   

                          l2               l3  rooms  bathrooms    price  \
6   Bs.As. G.B.A. Zona Oeste  Tres de Febrero    2.0        1.0      NaN   
10  Bs.As. G.B.A. Zona Norte            Tigre    4.0        4.0      NaN   
16           Capital Federal            Boedo    NaN        NaN  70000.0   
17  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  35000.0   
18  Bs.As. G.B.A. Zona Norte            Tigre    NaN        2.0  27000.0   

   currency property_type     operation_type  
6       NaN  Departamento           Alquiler  
10      NaN          Casa  Alquiler temporal  
16      ARS       Oficina           Alquiler  
17      ARS       Oficina           Alquiler  
18      ARS       Oficina           Alquiler

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (617747, 13)**

In [40]:
prop.shape

(617747, 13)

***Vamos a analizar la columna ad_type, que tipos de datos unicos tiene, si solo tiene un tipo de dato categorico eliminarla porque no aporta información.***

In [41]:
prop['ad_type'].unique()

array(['Propiedad'], dtype=object)

In [42]:
prop.drop(['ad_type'], axis=1, inplace=True)

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (617747, 12)**

In [43]:
prop.shape

(617747, 12)

# **AMBA**

***Vamos a analizar la columna l3, para eso primero vamos a filtrar solo los lugares que pertenecen al AMBA, los cuales son:***
    
```python
municipios = ['Ciudad Autónoma de Buenos Aires', 'Almirante Brown', 'Avellaneda', 'Berazategui', 'Esteban Echeverría', 'Ezeiza', 'Florencio Varela', 'General San Martín', 'Hurlingham', 'Ituzaingó', 'José C. Paz', 'La Matanza', 'Lanús', 'Lomas de Zamora', 'Malvinas Argentinas', 'Merlo', 'Moreno', 'Morón', 'Quilmes', 'San Fernando', 'San Isidro', 'San Miguel', 'Tigre', 'Tres de Febrero', 'Vicente López']
```


In [44]:
municipios = ['Ciudad Autónoma de Buenos Aires', 'Almirante Brown', 'Avellaneda', 'Berazategui', 'Esteban Echeverría', 'Ezeiza', 'Florencio Varela', 'General San Martín', 'Hurlingham', 'Ituzaingó', 'José C. Paz', 'La Matanza', 'Lanús', 'Lomas de Zamora', 'Malvinas Argentinas', 'Merlo', 'Moreno', 'Morón', 'Quilmes', 'San Fernando', 'San Isidro', 'San Miguel', 'Tigre', 'Tres de Febrero', 'Vicente López']

In [45]:
prop.loc[~(prop.l3.isin(municipios))].l3.unique()

array(['Boedo', 'Recoleta', 'Balvanera', 'San Nicolás',
       'General Rodríguez', 'Almagro', 'Palermo', 'Belgrano', 'La Plata',
       'San Cristobal', 'Barrio Norte', 'Constitución', 'Retiro',
       'Catalinas', 'Monserrat', 'Tribunales', 'Once', 'Caballito',
       'Villa Urquiza', 'Escobar', 'Pilar', 'Nuñez', 'Puerto Madero',
       'Floresta', 'Saavedra', 'Chacarita', 'Villa Crespo',
       'Villa Pueyrredón', 'Flores', 'Coghlan', 'Congreso',
       'Villa Devoto', 'Barracas', 'Colegiales', 'Villa del Parque',
       'San Vicente', 'José C Paz', 'Centro / Microcentro', 'San Telmo',
       'Monte Castro', 'Villa General Mitre', 'Parque Chacabuco', nan,
       'Abasto', 'Marcos Paz', 'Parque Chas', 'Villa Santa Rita',
       'Villa Lugano', 'Liniers', 'Mataderos', 'Las Cañitas', 'Paternal',
       'Villa Luro', 'Parque Patricios', 'Presidente Perón', 'Agronomía',
       'Villa Ortuzar', 'Villa Soldati', 'Parque Centenario', 'Cañuelas',
       'Parque Avellaneda', 'Pompeya', 'Boca'

**Sacar de la seleccion anterior los que pertenecen a Capital Federal y los valores NaN, luego ver cuales son los valores unicos de l3 para ver si queda alguno que pertenece a los municipios pero se escribio con alguna diferencia**

In [47]:
capital_federal = list(prop[(prop.l2 == "Capital Federal")].l3.unique())

In [48]:
capital_federal

['Boedo',
 'Recoleta',
 'Balvanera',
 'San Nicolás',
 'Almagro',
 'Palermo',
 'Belgrano',
 'San Cristobal',
 'Barrio Norte',
 'Constitución',
 'Retiro',
 'Catalinas',
 'Monserrat',
 'Tribunales',
 'Once',
 'Caballito',
 'Villa Urquiza',
 'Nuñez',
 'Puerto Madero',
 'Floresta',
 'Saavedra',
 'Chacarita',
 'Villa Crespo',
 'Villa Pueyrredón',
 'Flores',
 'Coghlan',
 'Congreso',
 'Villa Devoto',
 'Barracas',
 'Colegiales',
 'Villa del Parque',
 'Centro / Microcentro',
 'San Telmo',
 'Monte Castro',
 'Villa General Mitre',
 'Parque Chacabuco',
 nan,
 'Abasto',
 'Parque Chas',
 'Villa Santa Rita',
 'Villa Lugano',
 'Liniers',
 'Mataderos',
 'Las Cañitas',
 'Paternal',
 'Villa Luro',
 'Parque Patricios',
 'Agronomía',
 'Villa Ortuzar',
 'Villa Soldati',
 'Parque Centenario',
 'Parque Avellaneda',
 'Pompeya',
 'Boca',
 'Villa Real',
 'Versalles',
 'Velez Sarsfield',
 'Villa Riachuelo']

***Ahora, eliminar todos los registros que no pertenezcan al AMBA, ojo: no eliminar los que tengan valor NaN***

In [49]:
ambant = ['General Rodríguez',
 'La Plata',
 'Escobar',
 'Pilar',
 'San Vicente',
 'Marcos Paz',
 'Presidente Perón',
 'Cañuelas']

In [51]:
prop = prop.loc[~(prop.l3.isin(ambant))]

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (528442, 12)**

In [52]:
prop.shape

(528442, 12)

# **Latitud y Longitud**

***Vamos a analizar la calidad de los datos en las columnas de lat y lon. Buscaremos posibles outliers o datos incorrectos, para eso calcula para ambas columnas: el promedio, el maximo y el minimo.***

In [53]:
print(f"El promedio de la columna lat es: {prop['lat'].mean()}")
print(f"El máximo de la columna lat es: {prop['lat'].max()}")
print(f"El mínimo de la columna lat es: {prop['lat'].min()}")

El promedio de la columna lat es: -34.6013486009643
El máximo de la columna lat es: 42.4312952
El mínimo de la columna lat es: -53.788624


In [54]:
print(f"El promedio de la columna lon es: {prop['lon'].mean()}")
print(f"El máximo de la columna lon es: {prop['lon'].max()}")
print(f"El mínimo de la columna lon es: {prop['lon'].min()}")

El promedio de la columna lon es: -58.49057199897227
El máximo de la columna lon es: 58.703066
El mínimo de la columna lon es: -122.4552858412


In [55]:
round(prop.lat.describe(),2)

count    500380.00
mean        -34.60
std           0.43
min         -53.79
25%         -34.65
50%         -34.60
75%         -34.56
max          42.43
Name: lat, dtype: float64

In [56]:
round(prop.lon.describe(),2)

count    500341.00
mean        -58.49
std           0.45
min        -122.46
25%         -58.55
50%         -58.46
75%         -58.40
max          58.70
Name: lon, dtype: float64

***Evidentemente hay ciertos puntos que son incorrectos, sus coordenadas caen fuera del AMBA. Para eso buscamos cuales son los valores frontera para la latitud y longitud en el AMBA.***

***Estimamos un segmentos de latitudes de [-35, -34] y longitudes de [-59 a -58]***

***Eliminar todo lo que este fuera de esos rangos (no eliminar valores NaN)***

In [57]:
prop = prop[prop.lat.between(-35, -34) & prop.lon.between(-59, -58)]

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (499739, 12)**

In [58]:
prop.shape

(499739, 12)

***Volver a chequear estadisticos de lat y lon para ver si ya saque valores incorrectos***

In [60]:
print(f"El promedio de la columna lat es: {prop['lat'].mean()}")
print(f"El máximo de la columna lat es: {prop['lat'].max()}")
print(f"El mínimo de la columna lat es: {prop['lat'].min()}")
print("")
print(f"El promedio de la columna lon es: {prop['lon'].mean()}")
print(f"El máximo de la columna lon es: {prop['lon'].max()}")
print(f"El mínimo de la columna lon es: {prop['lon'].min()}")

El promedio de la columna lat es: -34.605433407124146
El máximo de la columna lat es: -34.0490256944
El mínimo de la columna lat es: -34.9881942684

El promedio de la columna lon es: -58.48363818980686
El máximo de la columna lon es: -58.021731
El mínimo de la columna lon es: -58.9818835258


In [61]:
round(prop.lat.describe(), 2)

count    499739.00
mean        -34.61
std           0.10
min         -34.99
25%         -34.65
50%         -34.60
75%         -34.56
max         -34.05
Name: lat, dtype: float64

In [62]:
round(prop.lon.describe(), 2)

count    499739.00
mean        -58.48
std           0.11
min         -58.98
25%         -58.55
50%         -58.46
75%         -58.40
max         -58.02
Name: lon, dtype: float64

***Eliminar los NaN de la columna l3***

In [63]:
prop['l3'].isna().value_counts()

False    499633
True        106
Name: l3, dtype: int64

In [64]:
prop.dropna(subset = ['l3'], inplace=True) 

In [65]:
prop['l3'].isna().value_counts()

False    499633
Name: l3, dtype: int64

***Volvamos a ver el porcentaje de nulos de cada columna***

In [66]:
(prop.isna().sum()*100/prop.shape[0]).sort_values(ascending = False)

rooms             19.392834
bathrooms         15.336257
currency           1.050971
price              0.939490
start_date         0.000000
end_date           0.000000
lat                0.000000
lon                0.000000
l2                 0.000000
l3                 0.000000
property_type      0.000000
operation_type     0.000000
dtype: float64

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (499633, 12)**

In [67]:
prop.shape

(499633, 12)

***Veamos los valores unicos de la columna property_type***

In [68]:
prop['property_type'].unique()

array(['Departamento', 'Casa', 'Oficina', 'Local comercial', 'Lote',
       'Cochera', 'PH', 'Otro', 'Depósito', 'Casa de campo'], dtype=object)

***La columna se ve bien, hacemos lo mismo con operation type***

In [69]:
prop['operation_type'].unique()

array(['Alquiler', 'Alquiler temporal', 'Venta'], dtype=object)

***La columna se ve bien. Nos quedan price, currency, bathrooms y rooms.***

***Vamos a currency, veamos los valores unicos de esta columna***

In [70]:
prop['currency'].unique()

array([nan, 'ARS', 'USD', 'COP'], dtype=object)

***Tenemos el peso argentino, el dolar, el peso colombiano y NaN. Veamos los registros con NaN.***

In [71]:
prop['currency'].isna().value_counts()

False    494382
True       5251
Name: currency, dtype: int64

In [72]:
prop[(prop.currency != "ARS") & 
     (prop.currency != "USD") &
     (prop.currency != "COP")]

start_date    end_date        lat        lon  \
6       2020-08-22  2020-09-04 -34.604195 -58.559827   
10      2020-08-22  2020-08-25 -34.439610 -58.629968   
649     2021-03-21  9999-12-31 -34.804675 -58.173498   
1140    2021-03-21  2021-04-23 -34.630972 -58.424762   
1228    2021-03-21  2021-03-22 -34.610553 -58.361774   
...            ...         ...        ...        ...   
997026  2020-12-04  2021-01-27 -34.648968 -58.577701   
997029  2020-12-04  2020-12-09 -34.761623 -58.207809   
997032  2020-12-04  2021-03-16 -34.919530 -58.479625   
997053  2020-12-04  2021-01-13 -34.611887 -58.359590   
997060  2020-12-04  2021-01-02 -34.644868 -58.469612   

                              l2                  l3  rooms  bathrooms  price  \
6       Bs.As. G.B.A. Zona Oeste     Tres de Febrero    2.0        1.0    NaN   
10      Bs.As. G.B.A. Zona Norte               Tigre    4.0        4.0    NaN   
649       Bs.As. G.B.A. Zona Sur         Berazategui    2.0        2.0    NaN   
1140             Capital Federal               Boedo    4.0        1.0    NaN   
1228             Capital Federal       Puerto Madero    3.0        2.0    NaN   
...                          ...                 ...    ...        ...    ...   
997026  Bs.As. G.B.A. Zona Oeste          La Matanza    4.0        2.0    NaN   
997029    Bs.As. G.B.A. Zona Sur         Berazategui   14.0        1.0    NaN   
997032    Bs.As. G.B.A. Zona Sur  Esteban Echeverría    5.0        4.0    NaN   
997053           Capital Federal       Puerto Madero    4.0        3.0    NaN   
997060           Capital Federal              Flores    9.0        3.0    NaN   

       currency property_type     operation_type  
6           NaN  Departamento           Alquiler  
10          NaN          Casa  Alquiler temporal  
649         NaN  Departamento              Venta  
1140        NaN  Departamento              Venta  
1228        NaN  Departamento              Venta  
...         ...           ...                ...  
997026      NaN          Casa              Venta  
997029      NaN          Casa              Venta  
997032      NaN          Casa              Venta  
997053      NaN  Departamento              Venta  
997060      NaN          Casa              Venta  

[5251 rows x 12 columns]

In [73]:
prop[(prop.currency != "ARS") & 
     (prop.currency != "USD") &
     (prop.currency != "COP")].currency.unique()

array([nan], dtype=object)

***Vemos que la cantidad de NaN en esta columna es del 1%, sin esta columna no podremos saber en que moneda esta el precio, por lo que tampoco sabremos el precio (el precio es la columna mas interesante de este dataset). Decidimos eliminar los registros con NaN de la columna currency.***

In [74]:
prop.dropna(subset = ['currency'], inplace=True)

In [75]:
prop['currency'].isna().value_counts()

False    494382
Name: currency, dtype: int64

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (494382, 12)**

In [77]:
prop.shape

(494382, 12)

***Volver a analizar el porcentaje de NaN de cada columna.***

In [78]:
(prop.isna().sum()*100/prop.shape[0]).sort_values(ascending = False)

rooms             19.570090
bathrooms         15.492676
start_date         0.000000
end_date           0.000000
lat                0.000000
lon                0.000000
l2                 0.000000
l3                 0.000000
price              0.000000
currency           0.000000
property_type      0.000000
operation_type     0.000000
dtype: float64

***Vemos que se eliminaron los NaN de price al eliminar los de currency, matamos 2 pajaros de un tiro.***

***Ahora si crear una nueva columna llamada precio con el valor de las propiedades en dolares. Utilizar estas cotizaciones***

1 dolar = $283 pesos argentinos

1 dolar = $4637 pesos colombianos

In [79]:
cond = [(prop['currency'] == "ARS"), 
        (prop['currency'] == "COP"), 
        (prop['currency'] == "USD")]

selec = [(prop['price'] / 283), 
         (prop['price'] / 4637), 
         (prop['price'])]

prop['precio'] = np.select(cond, selec)
prop

start_date    end_date        lat        lon  \
16      2020-08-22  2020-09-04 -34.633053 -58.419906   
17      2020-08-22  2020-09-04 -34.397401 -58.649185   
18      2020-08-22  2020-09-04 -34.397401 -58.649185   
19      2020-08-22  2020-09-04 -34.598112 -58.400784   
20      2020-08-22  2020-09-04 -34.604054 -58.393770   
...            ...         ...        ...        ...   
999993  2020-12-04  9999-12-31 -34.482248 -58.501989   
999994  2020-12-04  2021-01-22 -34.485005 -58.685923   
999996  2020-12-04  9999-12-31 -34.596587 -58.376893   
999997  2020-12-04  9999-12-31 -34.576445 -58.432603   
999999  2020-12-04  2020-12-21 -34.447524 -58.631552   

                              l2                   l3  rooms  bathrooms  \
16               Capital Federal                Boedo    NaN        NaN   
17      Bs.As. G.B.A. Zona Norte                Tigre    NaN        2.0   
18      Bs.As. G.B.A. Zona Norte                Tigre    NaN        2.0   
19               Capital Federal             Recoleta    NaN        3.0   
20               Capital Federal            Balvanera    NaN        1.0   
...                          ...                  ...    ...        ...   
999993  Bs.As. G.B.A. Zona Norte           San Isidro   11.0        4.0   
999994  Bs.As. G.B.A. Zona Norte  Malvinas Argentinas    NaN        7.0   
999996           Capital Federal               Retiro   12.0        5.0   
999997           Capital Federal              Palermo   14.0        2.0   
999999  Bs.As. G.B.A. Zona Norte                Tigre    5.0        4.0   

           price currency    property_type     operation_type         precio  
16       70000.0      ARS          Oficina           Alquiler     247.349823  
17       35000.0      ARS          Oficina           Alquiler     123.674912  
18       27000.0      ARS          Oficina           Alquiler      95.406360  
19       65000.0      ARS          Oficina           Alquiler     229.681979  
20       60000.0      ARS          Oficina           Alquiler     212.014134  
...          ...      ...              ...                ...            ...  
999993  690000.0      USD             Casa              Venta  690000.000000  
999994    8000.0      USD             Casa  Alquiler temporal    8000.000000  
999996    3700.0      USD  Local comercial           Alquiler    3700.000000  
999997  570000.0      USD  Local comercial              Venta  570000.000000  
999999    7000.0      USD             Casa  Alquiler temporal    7000.000000  

[494382 rows x 13 columns]

**PUNTO DE REVISION**

**A esta altura el dataframe tiene (499382, 13)**

In [80]:
prop.shape

(494382, 13)